cmd/gcp-controller-manager/dpwi/pods/eventhandler.go

```diff
@@ -0,0 +1,156 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package pods listens to and process pod events. If the KSA of the pod
+// has a verified non-empty GSA, it sends a node sync event.
+package pods
+
+import (
+	"fmt"
+	"time"
+
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/cloud-provider-gcp/cmd/gcp-controller-manager/dpwi/serviceaccounts"
+	"k8s.io/klog/v2"
+)
+
+const (
+	eventHandlerResyncPeriod = 30 * time.Minute
+)
+
+type eventHandler struct {
+	podIndexer cache.Indexer
+	hasSynced  func() bool
+	queue      workqueue.RateLimitingInterface
+	verifier   *serviceaccounts.Verifier
+	syncer     Syncer
+}
+
+type Syncer interface {
+	Enqueue(key interface{})
+}
+
+// NewEventHandler creates a new pod eventHandler.
+func NewEventHandler(informer coreinformers.PodInformer, verifier *serviceaccounts.Verifier, syncer Syncer) (*eventHandler, error) {
+	eh := &eventHandler{
+		podIndexer: informer.Informer().GetIndexer(),
+		hasSynced:  informer.Informer().HasSynced,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 1000*time.Second),
+		), "pod-event-queue"),
+		verifier: verifier,
+		syncer:   syncer,
+	}
+	informer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    eh.onPodAdd,
+		UpdateFunc: eh.onPodUpdate,
+		DeleteFunc: eh.onPodDelete,
+	}, eventHandlerResyncPeriod)
+	return eh, nil
+}
+
+func (eh *eventHandler) onPodAdd(obj interface{}) {
+	eh.Enqueue(obj)
+}
+
+func (eh *eventHandler) onPodUpdate(obj, oldObj interface{}) {
+	eh.Enqueue(obj)
+}
+
+func (eh *eventHandler) onPodDelete(obj interface{}) {
+	eh.Enqueue(obj)
+}
+
+// Enqueue puts a pod event into the pods workerqueue.
+func (eh *eventHandler) Enqueue(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.Errorf("internal error. Couldn't get key for Pod %+v: %v", obj, err)
+		return
+	}
+	eh.queue.Add(key)
+}
+
+// Run starts a number of workers to process pods events. The workers will
+// continue to run and process items from the workqueue until receiving stopCh.
+func (eh *eventHandler) Run(workers int, stopCh <-chan struct{}) {
+	for i := 0; i < workers; i++ {
+		go wait.Until(eh.work, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (eh *eventHandler) work() {
+	for eh.processNext() {
+	}
+}
+
+func (eh *eventHandler) processNext() bool {
+	key, quit := eh.queue.Get()
+	if quit {
+		return false
+	}
+	defer eh.queue.Done(key)
+
+	err := eh.process(key.(string))
```

Comment:
Similar comment as in the node syncer: each log line in this file should tell us a loop ID and the key being processed.